Value store v3.0

grokys opened the pull request:
## What does the pull request do?

This PR implements a new value store for `AvaloniaObject`. Previously, each property value was stored in a dictionary inside `ValueStore` which mapped `property` -> `value`, where `value` changed type depending on what was stored (single local value, single binding, multiple values of differing priorities). 

This PR changes `ValueStore` to store property values in a different manner;

- A dictionary mapping the property to an `EffectiveValue<T>` containing the effective value and base value + priorities
- Local values are stored directly to the `EffectiveValue<T>` for the property
- A collection of "Frames" ordered by priority. Frames are currently one of two types:
  - `ImmediateValueFrame`: values set by `SetValue` or `AddBinding` with non-LocalValue priority.
  - `StyleInstance`: a frame containing the `Setter`s for a style
- Each frame is a collection of `IValueEntry`s
- Inherited values work by storing a reference to the nearest ancestor which has inherited values set in its effective value collection

Some improvements the new value store enables:

- Frames can be added to multiple `ValueStore`s meaning that `Style`s which don't have activators (and `ControlTheme`s) can be instanced just once instead of needing to be instanced on each `AvaloniaObject` separately.
- `Setter` is an `IValueEntry` meaning that it can be stored directly a frame unless it needs to be explicitly instanced on each control
- Frames can be added and removed separately, meaning that a `ControlTheme` can be changed without affecting the rest of the styles applied to it (this isn't actually implemented yet - will require a change to our `BindingPriority`s in order to distinguish between control themes and styles)
- We don't need to traverse the whole tree to read an inherited value - only the ancestors in the tree which set inherited values
- Direct property bindings can be stored in the same collection as `LocalValue` styled property bindings, removing the need for 2 lists per `AvaloniaObject`
- Uses a minimum amount of virtual methods and generic methods as these can slow things down a lot (virtual methods at runtime, generic methods at JIT time)

## Benchmarks

```
| Slower                                                                          | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| ------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| Avalonia.Benchmarks.Base.AvaloniaObject_GetValue.GetDefaultValues               |      1.17 |         14114.19 |         16473.92 |         |
| Avalonia.Benchmarks.Styling.ResourceBenchmarks.FindNotExistingResource          |      1.12 |        809786.47 |        909711.54 |         |
| Avalonia.Benchmarks.Base.AvaloniaObjectInitializationBenchmark.InitializeButton |      1.12 |           217.07 |           243.48 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| Avalonia.Benchmarks.Base.AvaloniaObject_GetValueInherited.GetInheritedValues(Dep |     18.94 |        393126.81 |         20757.63 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_GetValueInherited.GetInheritedValues(Dep |      9.65 |        199979.01 |         20728.13 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_GetValueInherited.GetInheritedValues(Dep |      5.54 |        117037.06 |         21110.74 |         |
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 50,  |      4.22 |         24211.55 |          5732.49 |         |
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 50,  |      4.18 |         24524.41 |          5869.84 |         |
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 50,  |      3.50 |         26346.97 |          7524.48 |         |
| Avalonia.Benchmarks.Styling.Style_ClassSelector.Apply_Toggle                     |      3.30 |          1618.19 |           489.66 |         |
| Avalonia.Benchmarks.Styling.Style_ClassSelector.Apply                            |      3.18 |           866.49 |           272.54 |         |
| Avalonia.Benchmarks.Styling.Style_ClassSelector.Apply_Detach                     |      3.02 |          1392.40 |           461.58 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_GetValue.Get_Local_Values                |      2.81 |         39732.18 |         14129.47 |         |
| Avalonia.Benchmarks.Base.AvaloniaObjectBenchmark.BindIntProperty                 |      2.75 |         14234.11 |          5171.49 |         |
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Set_Coerced_Int_Property_Local |      2.62 |         17512.68 |          6680.21 |         |
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 5, N |      2.60 |          2819.56 |          1085.22 |         |
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Set_Int_Property_Multiple_Prio |      2.54 |        254065.91 |         99880.78 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_GetValue.Get_Local_Values_With_Style_Val |      2.52 |         35301.84 |         13988.89 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_SetValue.Set_Local_Values_With_Style_Val |      2.51 |        212792.09 |         84756.34 |         |
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 5, N |      2.40 |          3016.59 |          1256.49 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_GetValueInherited.GetInheritedValues(Dep |      2.22 |         47551.68 |         21380.50 |         |
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Set_Int_Property_LocalValue    |      2.13 |         13529.07 |          6364.65 |         |
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Set_Validated_Int_Property_Loc |      2.05 |         13719.47 |          6681.29 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_Binding.Fire_LocalValue_Bindings_With_St |      2.05 |        211422.13 |        103336.03 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_GetValueInherited.GetInheritedValues(Dep |      2.05 |         41533.68 |         20306.47 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_GetValueInherited.GetInheritedValues(Dep |      2.01 |         42798.10 |         21332.83 |         |
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Set_Int_Property_TemplatedPare |      1.88 |         38559.38 |         20458.80 |         |
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Bind_Int_Property_Multiple_Pri |      1.84 |         50798.83 |         27596.72 |         |
| Avalonia.Benchmarks.Rendering.ShapeRendering.Render_Line_WithFill                |      1.78 |            83.03 |            46.72 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_SetValue.SetValues                       |      1.66 |        140613.59 |         84628.17 |         |
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 5, N |      1.65 |          4733.56 |          2863.75 |         |
| Avalonia.Benchmarks.Base.AvaloniaObjectBenchmark.ClearAndSetIntProperty          |      1.63 |           262.61 |           161.28 |         |
| Avalonia.Benchmarks.Styling.Style_NonActive.Toggle_NonActive_Style_Activation    |      1.59 |         50500.13 |         31824.36 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_Binding.Setup_Dispose_LocalValue_Binding |      1.45 |        348959.30 |        240953.58 |         |
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 1, N |      1.43 |           917.89 |           640.52 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_GetObservable.PropertyChangedSubscriptio |      1.42 |        182705.85 |        129079.87 |         |
| Avalonia.Benchmarks.Base.DirectPropertyBenchmark.SetAndRaiseSimple               |      1.37 |          3833.04 |          2800.24 |         |
| Avalonia.Benchmarks.Base.DirectPropertyBenchmark.SetAndRaiseOriginal             |      1.36 |          3631.91 |          2664.83 |         |
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 1, N |      1.35 |          1097.20 |           811.80 |         |
| Avalonia.Benchmarks.Rendering.ShapeRendering.Render_Line_WithFillAndStroke       |      1.35 |           322.64 |           239.44 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_GetObservable.GetObservables             |      1.35 |        408508.36 |        303470.05 |         |
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Bind_Int_Property_LocalValue   |      1.33 |          7343.93 |          5529.49 |         |
| Avalonia.Benchmarks.Visuals.VisualAffectsRenderBenchmarks.SetPropertyThatAffects |      1.27 |           455.90 |           359.53 |         |
| Avalonia.Benchmarks.Themes.FluentBenchmark.RepeatButton                          |      1.26 |        208442.03 |        165051.30 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_Construct.Construct_And_Set_Values       |      1.24 |          2008.03 |          1618.78 |         |
| Avalonia.Benchmarks.Data.BindingsBenchmark.UpdateTwoWayBinding_Via_Binding       |      1.24 |        146089.96 |        117788.46 |         |
| Avalonia.Benchmarks.Layout.ControlsBenchmark.CreateButton                        |      1.21 |        127293.46 |        105071.62 |         |
| Avalonia.Benchmarks.Rendering.ShapeRendering.Render_Line_WithStroke              |      1.20 |           305.47 |           254.36 |         |
| Avalonia.Benchmarks.Layout.ControlsBenchmark.CreateCalendarWithLoaded            |      1.20 |      17490624.00 |      14626879.00 |         |
| Avalonia.Benchmarks.Layout.ControlsBenchmark.CreateCalendar                      |      1.19 |      17434331.00 |      14619440.00 |         |
| Avalonia.Benchmarks.Styling.StyleAttachBenchmark.AttachTextBoxStyles             |      1.14 |          1386.99 |          1214.21 |         |
| Avalonia.Benchmarks.Base.AvaloniaObject_Binding.Fire_LocalValue_Bindings         |      1.14 |        121065.15 |        106660.97 |         |
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 1, N |      1.12 |          2726.70 |          2436.82 |         |

| MoreAllocations                                                                  | diff/base | Base Allocated (B) | Diff Allocated (B)|
| -------------------------------------------------------------------------------- | ---------:| ------------------:| ------------------:|
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Set_Int_Property_Multiple_Prio |      1.51 |           69392.00 |          104600.00|
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Set_Int_Property_TemplatedPare |      1.43 |           20800.00 |           29672.00|
| Avalonia.Benchmarks.Styling.Style_Activation.Toggle_Style_Activation_Via_Class   |      1.40 |           16000.00 |           22400.00|
| Avalonia.Benchmarks.Base.AvaloniaObjectInitializationBenchmark.InitializeButton  |      1.09 |             928.00 |            1016.00|
| Avalonia.Benchmarks.Base.AvaloniaObject_Construct.Construct_And_Set_Values       |      1.07 |            2408.00 |            2584.00|
| Avalonia.Benchmarks.Data.BindingsBenchmark.TwoWayBinding_Via_Binding             |      1.04 |            2867.00 |            2981.00|
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Bind_Int_Property_LocalValue   |      1.03 |            6768.00 |            7000.00|
| Avalonia.Benchmarks.Visuals.Media.PathMarkupParserTests.Parse_Large_Path         |      1.03 |            3608.00 |            3712.00|
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Set_Int_Property_LocalValue    |      1.01 |            6616.00 |            6712.00|
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Set_Validated_Int_Property_Loc |      1.01 |            6616.00 |            6712.00|

| LessAllocations                                                                  | base/diff | Base Allocated (B) | Diff Allocated (B)|
| -------------------------------------------------------------------------------- | ---------:| ------------------:| ------------------:|
| Avalonia.Benchmarks.Styling.Style_NonActive.Toggle_NonActive_Style_Activation    |      0.00 |           16000.00 |               0.00|
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 50,  |      0.11 |           20384.00 |            2320.00|
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 50,  |      0.11 |           20384.00 |            2320.00|
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 50,  |      0.11 |           20384.00 |            2320.00|
| Avalonia.Benchmarks.Base.AvaloniaObject_Binding.Setup_Dispose_LocalValue_Binding |      0.25 |          459201.00 |          112800.00|
| Avalonia.Benchmarks.Styling.StyleAttachBenchmark.AttachTextBoxStyles             |      0.37 |             736.00 |             272.00|
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 5, N |      0.41 |            3328.00 |            1352.00|
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 5, N |      0.41 |            3328.00 |            1352.00|
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 5, N |      0.41 |            3328.00 |            1352.00|
| Avalonia.Benchmarks.Traversal.VisualTreeTraversal.FindCommonVisualAncestor       |      0.42 |           12688.00 |            5312.00|
| Avalonia.Benchmarks.Base.AvaloniaObjectBenchmark.BindIntProperty                 |      0.42 |           15528.00 |            6560.00|
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Bind_Int_Property_Multiple_Pri |      0.43 |           39904.00 |           17216.00|
| Avalonia.Benchmarks.Styling.Style_ClassSelector.Apply_Toggle                     |      0.47 |             672.00 |             313.00|
| Avalonia.Benchmarks.Base.AvaloniaObject_SetValue.Set_Local_Values_With_Style_Val |      0.48 |          168000.00 |           80000.00|
| Avalonia.Benchmarks.Base.AvaloniaObject_Binding.Fire_LocalValue_Bindings_With_St |      0.51 |          182457.00 |           92488.00|
| Avalonia.Benchmarks.Base.StyledPropertyBenchmarks.Set_Coerced_Int_Property_Local |      0.51 |           13104.00 |            6744.00|
| Avalonia.Benchmarks.Styling.Style_ClassSelector.Apply_Detach                     |      0.52 |             589.00 |             305.00|
| Avalonia.Benchmarks.Styling.Style_ClassSelector.Apply                            |      0.60 |             508.00 |             305.00|
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 1, N |      0.78 |            1624.00 |            1264.00|
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 1, N |      0.78 |            1624.00 |            1264.00|
| Avalonia.Benchmarks.Styling.Style_Apply.Apply_Simple_Styles(MatchingStyles: 1, N |      0.78 |            1624.00 |            1264.00|
| Avalonia.Benchmarks.Data.BindingsBenchmark.UpdateTwoWayBinding_Via_Binding       |      0.78 |           34812.00 |           27103.00|
| Avalonia.Benchmarks.Themes.FluentBenchmark.RepeatButton                          |      0.84 |           38618.00 |           32363.00|
| Avalonia.Benchmarks.Styling.SelectorBenchmark.ClassSelector_NoMatch              |      0.86 |              56.00 |              48.00|
| Avalonia.Benchmarks.Styling.SelectorBenchmark.ClassSelector_Match                |      0.86 |              56.00 |              48.00|
| Avalonia.Benchmarks.Layout.ControlsBenchmark.CreateButton                        |      0.86 |           31010.00 |           26651.00|
| Avalonia.Benchmarks.Layout.ControlsBenchmark.CreateCalendar                      |      0.88 |         4240880.00 |         3717504.00|
| Avalonia.Benchmarks.Layout.ControlsBenchmark.CreateCalendarWithLoaded            |      0.88 |         4267224.00 |         3743848.00|
| Avalonia.Benchmarks.Base.AvaloniaObjectBenchmark.ClearAndSetIntProperty          |      0.88 |             208.00 |             184.00|
| Avalonia.Benchmarks.Layout.ControlsBenchmark.CreateTextBox                       |      0.90 |          235048.00 |          212624.00|
| Avalonia.Benchmarks.Base.AvaloniaObject_SetValue.SetValues                       |      0.91 |           88000.00 |           80000.00|
| Avalonia.Benchmarks.Themes.ThemeBenchmark.InitSimpleTheme(mode: Dark)            |      0.97 |          235682.00 |          228717.00|
| Avalonia.Benchmarks.Base.AvaloniaObject_Binding.Fire_LocalValue_Bindings         |      0.97 |           95472.00 |           92890.00|
| Avalonia.Benchmarks.Themes.ThemeBenchmark.InitFluentTheme(mode: Dark)            |      0.99 |          544204.00 |          536954.00|
| Avalonia.Benchmarks.Themes.ThemeBenchmark.InitFluentTheme(mode: Light)           |      0.99 |          543836.00 |          536707.00|
```

Of these, `InitializeButton` is only 26ns slower which is on the limit of what can be measured.

The nearly 19x improvement in reading inherited values should really make a difference in some applications.

Memory benchmarks have now been added. Most of increases in memory are because of the particular scenarios the benchmarks are testing and should be offset by the decreases in a real app.

# Breaking Changes

Removes `BeginBatchUpdate`/`EndBatchUpdate` and replaces it with internal `BeginStyling`/`EndStyling` as the current batch update system introduced in 0.10.1 isn't zero cost, whereas the more limited `BeginStyling`/`EndStyling` is.

I want to re-think and maybe re-implement batch updates separately from this PR, at least for initialization, but if anyone is using them please let me know how/where.

Depends on #8571

MarchingCube committed:
Refactor value storage to reduce overhead and memory allocations.

grokys committed:
Initial refactor of AvaloniaObject value store.

Most (but not all) tests passing, all features mostly implemented exception coercion.

grokys committed:
Don't add duplicate frames.

grokys committed:
Use custom binary search in ValueStore.

grokys committed:
Refactored inheritance.

To remove `InheritanceFrame` - it was unneeded, we can just point to the nearest ancestor value store with inherited values.

grokys committed:
Implemented coercion.

grokys committed:
Reinstated test.

grokys committed:
Stop assert dialogs popping up in netfx tests.

grokys committed:
Don't produce value when reading (Has)Value.

grokys committed:
Don't start values unnecessarily.

grokys committed:
Fix DataContext notifications.

grokys committed:
Tweak naming.

grokys committed:
Remove unused field.

grokys committed:
Fix missing logging.

grokys committed:
Unwrap exceptions before logging.

grokys committed:
Fix a couple of inheritance bugs.

With associated unit tests.

grokys committed:
Refactored style activators.

- Always evaluate the active state from current information, don't rely on subscriptions to fire as the current state may not be up-to-date
- Don't notify the `IStyleActivatorSink` of a change immediately on subscription

grokys committed:
Remove IValueFrame.

And use the concrete `ValueFrame` (renamed from `ValueFrameBase`) as this was the only implementation of the interface.

grokys committed:
Make ValueFrame.Priority non-virtual.

grokys committed:
Use shared StyleInstance if possible.

If a `StyleBase` has no activator, and its `Setter`s don't need to be separately instanced on the control, then we can share a `StyleInstance` between controls.

This causes failing tests because `TopLevel` applies styles in its constructor, but doesn't set `_styled` to true, meaning that styles get applied twice. This issue will need to be addressed separately.

grokys committed:
Stop assert dialogs popping up in leak tests.

grokys committed:
Remove assertion.

Due to #8549.

grokys committed:
Remove unused method.

grokys committed:
Merge branch 'master' into value-storage-perf-refactor

grokys committed:
WIP: Tweaking benchmarks.

grokys committed:
WIP: Removed initializing, use try pattern.

grokys committed:
WIP: Speed up benchmarks.

grokys committed:
Store property instead of id in Entry.

grokys committed:
Conform better to dictionary API.

And rename collection to indicate this.

grokys committed:
Add null check.

grokys committed:
Remove some null forgiving operators.

grokys committed:
Added ContainsKey.

grokys committed:
Added tests for AvaloniaPropertyDictionary.

And fixed a couple of bugs.

grokys committed:
Renamed/moved benchmark.

grokys committed:
Remove unneeded change.

grokys committed:
Added Clear method.

grokys committed:
Merge branch 'refactor/avaloniapropertystore' into refactor/new-value-store

grokys committed:
Use new AvaloniaPropertyDictionary.

grokys committed:
Make ValueFrame entries ordered.

grokys committed:
Throw for duplicate setters in styles.

And remove the duplicate setter that we have.

grokys committed:
Merge branch 'master' into refactor/new-value-store

grokys committed:
Merge branch 'master' into refactor/avaloniapropertystore

grokys committed:
Flush dispatcher queue in control benchmarks.

grokys committed:
Merge branch 'refactor/avaloniapropertystore' into refactor/new-value-store

grokys committed:
Detach control theme from template children.

grokys committed:
Merge branch 'master' into refactor/new-value-store

grokys committed:
Merge branch 'master' into refactor/new-value-store

grokys committed:
Don't throw trying to unsub if not subscribed.

This can happen when re-evaluating the value store if a style is detached before its frame is evaluated.

grokys committed:
Fix evaluation in presence of multiple animation frames.

grokys committed:
Store direct property bindings in ValueStore.

We already have a place to store them in the `_localValueBindings` list, so use that rather than having a separate `_directBindings` list in `AvaloniaObject`.

grokys committed:
Special-case frames with one entry.

Seems it's slightly faster to re-evaluate the single value in this case.

grokys committed:
Slightly optimized ReevaluateEffectiveValue.

grokys committed:
Remove unused code.

danipen commented:
@grokys there is a typo in the PR body...

> Depends on #8600

8600 is this PR itself.

grokys committed:
Don't set owner for shared value frames.

Should fix leak tests.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0022508-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys commented:
Oops, nice catch @danipen - fixed.

robloo commented:
There are some really interesting ideas here, nice work!

Sorien commented on tests/Avalonia.Benchmarks/Avalonia.Benchmarks.csproj:
```diff
@@ -5,6 +5,10 @@
   </PropertyGroup>
+  <Import Project="..\..\build\SharedVersion.props" />
+
+  <Import Project="..\..\build\SharedVersion.props" />
```
duplicate tag?

grokys replied:
Huh! How did that get there? fixed.

grokys committed:
Remove duplicate import.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0022570-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Merge branch 'master' into refactor/new-value-store

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0022831-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

Commit:
Merge branch 'master' into refactor/new-value-store

danwalmsley commented:
@grokys happened to notice that the indeterminate progress bar on control catalog stops working with this pr.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0023061-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Merge branch 'master' into refactor/new-value-store

grokys committed:
Added failing test for TwoWay bindings.

grokys committed:
Added more failing tests for desired behavior.

We should be unsubscribing from bindings when a higher priority binding or value is added.

grokys committed:
Remove hack from PropertySetterBindingInstance.

grokys committed:
Unsubscribe inactive bindings.

grokys committed:
More failing tests for two-way bindings.

grokys committed:
Refactor binding IValueEntry into base class.

Too many gotchas in implementing binding entries, we'll take the performance hit of a few virtual method calls.

grokys committed:
Fix inactive two-way setter bindings.

Make two-way bindings in setters not write to source when inactive.

grokys committed:
Skip some of the two-way binding tests.

The value store alone cannot fix these tests, as writing back to source in two-way bindings is implemented at a higher level than the value store. We'll need to refactor the way two-way bindings work in a separate PR.

grokys committed:
Added failing animation test.

grokys committed:
Attach animations in styles.

grokys committed:
Expand nesting selectors in Style.ToString.

grokys committed:
Add additional failing test.

grokys committed:
Re-evaluate all properties if necessary.

When re-evaluating a single property, if the frame activation state changes then we need to re-evaluate all properties.

grokys committed:
Expose activation state via a method.

Previously, the `IsActive` property caused the activator state to be re-evaluated, which meant that when the _debugger_ read it, it was re-evaluated, making debugging difficult.

grokys committed:
Remove tags from style activators.

They're no longer needed.

grokys committed:
Added another failing test.

grokys committed:
Re-evaluate all properties if necessary.

When re-evaluating a single property, if the frame activation state  changes when unsubscribing from inactive frames then we need to re-evaluate all properties.

grokys committed:
Make test fail again.

grokys committed:
Re-evaluate all properties if necessary.

Even if the frame's activation state has changed to false.

grokys committed:
Include inherited/default values in diagnostic.

grokys committed:
Added failing inheritance test.

grokys committed:
Raise correct inheritance value change.

When raising `PropertyChanged` due to setting the inheritance parent, use the closest value in the inheritance tree rather than the furthest.

grokys committed:
Don't use Microsoft.Reactive.Testing here.

It's unsuitable for our needs as there's no way to make an observable fire on subscribe.

grokys committed:
Make EffectiveValue responsible for unsubscribing.

grokys committed:
Fix bug in ExpressionObserver.

Ensure `rootGetter` is evaluated each time the observer is initialized; plus failing and now passing test.

grokys committed:
Simplify unset effective value removal.

grokys committed:
Correctly unsubscribe animation value entries.

grokys committed:
Go direct to ValueStore for untyped GetValue.

Avoids a virtual call in the case of styled properties.

grokys committed:
Removed unused method.

grokys committed:
Remove unused variables/parameters.

grokys committed:
Remove unused methods and add docs.

grokys committed:
Merge branch 'master' into refactor/new-value-store

grokys committed:
Merge branch 'master' into refactor/new-value-store

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0025391-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Don't allocate _entries until >1 value.

grokys committed:
Remove unneeded interface.

grokys committed:
Remove unneeded method.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0025427-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Remove unused code.

grokys committed:
Added a few more coercion tests (1 failing).

grokys committed:
Raise base value changed event on coercion.

grokys commented:
Ok, I think this is ready for review/test/merge.

maxkatz6 commented on src/Avalonia.Base/AvaloniaObject.cs:
```diff
@@ -269,18 +242,10 @@ public T GetValue<T>(DirectPropertyBase<T> property)
-
-            return default;
+            _ = property ?? throw new ArgumentNullException(nameof(property));
+            return _values.GetBaseValue(property);
```
Was VerifyAccess missed here?

maxkatz6 commented on src/Avalonia.Base/AvaloniaObject.cs:
```diff
@@ -675,25 +608,55 @@ protected virtual void OnPropertyChanged(AvaloniaPropertyChangedEventArgs change
+        /// <param name="oldValue">The old property value.</param>
+        /// <param name="newValue">The new property value.</param>
+        /// <param name="priority">The priority of the binding that produced the value.</param>
+        protected void RaisePropertyChanged<T>(
+            DirectPropertyBase<T> property,
+            Optional<T> oldValue,
+            BindingValue<T> newValue,
+            BindingPriority priority = BindingPriority.LocalValue)
+        {
+            RaisePropertyChanged(property, oldValue, newValue, priority, true);
+        }
```
Haven't seen this API before. Not sure how it should work, as this notification can be "detached" from the actual property value. As developer can pass fake old and new value properties, which will raise notification, but actual value wasn't changed.
Was it probably intended to be "protected internal"?

grokys replied:
Will change this to `private protected`.

maxkatz6 commented on src/Avalonia.Base/CollectionPolyfills.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace Avalonia
+{
+    internal static class CollectionPolyfills
```
Can be moved to Avalonia.Base\Compatibility folder, as we have some similar helpers there already

maxkatz6 commented on src/Avalonia.Base/Data/BindingPriority.cs:
```diff
@@ -35,6 +35,11 @@ public enum BindingPriority
+        /// <summary>
+        /// The value is inherited from an ancestor element.
+        /// </summary>
+        Inherited,
```
Isn't it internal only state? As I see so far, developer can't get priority of the property value from the public API, and calling SetValue with this makes less sense.

grokys replied:
Not really, because it is exposed on the property change event args, where it's a valid value. Also we already have `Unset` priority in this enum. We should probably add a check to ensure that no-one tries to set/bind properties with these enum values though.

grokys replied:
Done.

maxkatz6 commented on src/Avalonia.Base/PropertyStore/DirectBindingObserver.cs:
```diff
@@ -0,0 +1,76 @@
+                var instance = _owner.Owner;
+                var property = Property;
+                var newValue = value;
+                Dispatcher.UIThread.Post(() => instance.SetDirectValueUnchecked(property, newValue));
```
I think we have Post method overload that accepts an object parameter.
Something like:
```
Dispatcher.UIThread.Post(static tuple =>
{
var (instance, property, newValue) = ValueTuple<InstanceType, DirectPropertyBase<T>, T>tuple;
instance.SetDirectValueUnchecked(property, newValue);
}, (_owner.Owner, Property, value));
```
Which will eliminate closures completely.

Also, that's an ugly way to do it, but hey.

maxkatz6 replied:
If this code isn't too ugly, it can be used in multiple places of this PR and later in the codebase in overall.

grokys replied:
This code is unchanged from the [equivalent of this class on `master`](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/AvaloniaObject.cs#L945-L961). If we decide to do this in a different way, we can do it as a separate PR IMO.

MarchingCube replied:
Yeah I wrote it quite some time ago before we had parametrized `Post`, but even now that API won't allow for really cheap job posting.

MarchingCube commented on src/Avalonia.Base/Utilities/AvaloniaPropertyDictionary.cs:
```diff
@@ -0,0 +1,369 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using static Avalonia.Rendering.Composition.Animations.PropertySetSnapshot;
```
Can be removed.

MarchingCube reviewed (commented):
I went through some of your changes and used control catalog for a bit. So far I have found two app crashes:
- resizing the window to make sidebar collapse and then expand, then close the app - it will crash in property value store
- go to "Native Embed" page and click on "Tooltip", usually it will crash with out of bounds exception

Edit: It seems that "TransitioningContentControl" page has issues now, content seems to transition twice on page open

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0025534-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

MarchingCube commented on src/Avalonia.Base/AvaloniaObject.cs:
```diff
@@ -19,19 +18,18 @@ namespace Avalonia
         private List<AvaloniaObject>? _inheritanceChildren;
-        private ValueStore? _values;
-        private bool _batchUpdate;
+        private ValueStore _values;
```
Can be readonly.

maxkatz6 commented on src/Avalonia.Base/Styling/StyleInstance.cs:
```diff
@@ -1,137 +1,103 @@
         }
 
-        private void ActivatorChanged(bool value)
+        public new void MakeShared() => base.MakeShared();
```
Is it a debug leftover?

grokys replied:
Nope, base implementation is `protected`

maxkatz6 commented on src/Avalonia.Base/Styling/Activators/AndActivator.cs:
```diff
@@ -11,47 +9,45 @@ namespace Avalonia.Styling.Activators
+            for (var i = 0; i < count; ++i)
             {
-                PublishNext(_flags == _mask);
+                if (_sources[i].GetIsActive())
```
I see there is a difference from old code, that now we evaluate is active for each child activator, even if AndActivator.EvaluateIsActive is triggered always only on single child activator changed.
I assume it's not a big problem, as all our activators aren't expensive to recalculate is-active. It might change if we implement ":has()" selector tho.

grokys replied:
Yeah, there was a problem with how we did it before, in that it was susceptible to event ordering, in that the event that triggers a re-evaluation could be triggered _after_ the call to get the active state. This happens because of the way values are re-evaluated now, and that activators now apply to frames instead of individual properties. I'll try to leave an example of what happens here tomorrow.

grokys committed:
Make IsEvaluating re-entrancy friendly.

grokys committed:
Added failing transition test.

grokys committed:
Prevent exception when removing value.

Calling `DisposeAndRaiseUnset` can cause a nested reevaluation of property values.

grokys committed:
Add VerifyAccess check to GetBaseValue.

grokys committed:
This method is not a public API.

grokys committed:
Move collection polyfills to Compatibility dir.

And tweak to match existing code there.

grokys committed:
Make fields readonly.

And remove a `using` that crept in unexpectedly.

grokys commented on src/Avalonia.Base/Data/BindingValue.cs:
```diff
@@ -230,19 +231,105 @@ public BindingValue<T> WithValue(T value)
+            return new BindingValue<T>(type, v, error);
+        }
+
+        public static bool operator !=(BindingValue<T> x, Optional<T> y)
```
Actually I don't think these are used in the end. Will remove.

grokys committed:
Remove unused operators.

grokys committed:
Make field readonly.

grokys committed:
Remove empty frames.

In doing so, merged `ValueStore.OnBindingCompleted` into `OnValueEntryRemoved` as they were doing the same thing.

grokys committed:
Signal completion even when no observers.

grokys commented:
> resizing the window to make sidebar collapse and then expand, then close the app - it will crash in property value store

Fixed. Was caused by nested property invalidations.

> go to "Native Embed" page and click on "Tooltip", usually it will crash with out of bounds exception

Fixed, same issue as above.

> Edit: It seems that "TransitioningContentControl" page has issues now, content seems to transition twice on page open

Fixed, `SingleSubscriberObservableBase` didn't signal completion if an observable was completed with no observers, and in this new value store we unsubscribe from non-active bindings.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0025602-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Validate priority for set/bind.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0025609-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

jmacato committed:
Merge branch 'master' into refactor/new-value-store

MarchingCube commented on src/Avalonia.Base/StyledElement.cs:
```diff
@@ -830,59 +835,25 @@ private void ClearLogicalParent(IList children)
         }
 
-        private void DetachStyles()
+        private void DetachStyles(IReadOnlyList<StyleBase>? styles = null)
```
Based on usage here and how this list is built - maybe it makes sense to instead build a `HashSet`? This would allow for cheap lookup.

grokys replied:
Yeah probably, though this method comes from the code before this PR (i.e. there was already a `DetachStyles(IReadOnlyList<StyleBase>? styles)` method). Detaching styles will be refactored in future anyway because after this PR we should be able to remove individual frames.

I'd rather leave it as-is in this PR as it's not really related to the value store.

MarchingCube commented on src/Avalonia.Base/StyledElement.cs:
```diff
@@ -930,38 +901,24 @@ private void DetachStylesFromThisAndDescendents(IReadOnlyList<IStyle> styles)
-                result.Add(style);
-                RecurseStyles(style.Children, result);
+                var s = styles[i];
+                if (s is StyleBase style)
```
Maybe it would make sense to add a method to `IStyle` instead of using reflection data (type id).

grokys replied:
I need to refactor how styles are applied anyway, this was the simplest way to make this code work currently.

MarchingCube commented on src/Avalonia.Base/Styling/Setter.cs:
```diff
@@ -57,16 +57,72 @@ public Setter(AvaloniaProperty property, object value)
+                throw new InvalidOperationException(
+                    $"Cannot set direct property '{Property}' in '{instance.Source}' because the style has an activator.");
+
+            if (Value is IBinding binding)
```
Switch statement here would be easier for the eyes :)

grokys replied:
Yeah I did actually try out a switch statement here but it didn't look as clean to me:

```
            return Value switch
            {
                IBinding binding => SetBinding((StyleInstance)instance, ao, binding),
                ITemplate template when !typeof(ITemplate).IsAssignableFrom(Property.PropertyType) =>
                    new PropertySetterTemplateInstance(Property, template),
                _ when !Property.IsValidValue(Value) =>
                    throw new InvalidCastException($"Setter value '{Value}' is not a valid value for property '{Property}'."),
                _ when Property.IsDirect => SetDirectValue(target),
                _ => this,
            };
```

Same number of lines, but reads less clear than the `if` to me, mainly because it's testing a few different things beyond just `Value`.

MarchingCube commented on src/Avalonia.Base/Styling/StyleBase.cs:
```diff
@@ -80,11 +82,46 @@ public bool TryGetResource(object key, out object? result)
-            target.StyleApplied(instance);
-            instance.Start();
+            if (target is not AvaloniaObject ao)
+                throw new InvalidOperationException("Styles can only be applied to AvaloniaObjects.");
```
Due to GH formatting this almost looks like as it happens all the time. Maybe adding braces here wouln't hurt.

grokys replied:
Not sure what you mean here?

MarchingCube reviewed (commented):
Also after main changes are done - it would be worth it and go over added classes and seal everything to help devirtualization.

MarchingCube commented on src/Avalonia.Base/PropertyStore/EffectiveValue`1.cs:
```diff
@@ -0,0 +1,260 @@
+        {
+            Debug.Assert(priority != BindingPriority.LocalValue);
+            UpdateValueEntry(value, priority);
+            SetAndRaiseCore(owner, (StyledPropertyBase<T>)value.Property, (T)value.GetValue()!, priority);
```
This `(T)value.GetValue()!` is  boxing all the time. Shows up as a major GC point when profiling in larger apps.

grokys replied:
Committed a fix for this.

grokys committed:
Merge branch 'master' into refactor/new-value-store

grokys committed:
Avoid boxing in EffectiveValue<T>.SetAndRaise.

Added `IValueEntry<T>` interface back in and use that if present to get the value from the entry.

grokys commented on src/Avalonia.Base/PropertyStore/ValueStore.cs:
```diff
@@ -0,0 +1,964 @@
+                        {
+                            current = property.CreateEffectiveValue(Owner);
+                            AddEffectiveValue(property, current);
+                            current.SetAndRaise(this, entry, priority);
```
Hmm, good idea actually.

grokys replied:
Actually in combination with your idea below to cache property metadata, might not improve anything.

grokys replied:
I took your other suggestion in the end.

grokys committed:
Cache property metadata in effective value.

Decreases the number of metadata lookups needed. Previously we were doing two lookups on construction now we're only doing one.

grokys commented on src/Avalonia.Base/PropertyStore/EffectiveValue`1.cs:
```diff
@@ -0,0 +1,260 @@
+            Priority = priority;
+            BasePriority = BindingPriority.Unset;
+
+            if (property.HasCoercion &&
```
Ok, implemented this. I also use the cached metadata when inherited values change now, which may have different behavior in the case of inherited values which have overridden default values. I think we need to disallow overriding default values for inherited properties because the semantics aren't clear anyway.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0025922-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Merge branch 'master' into refactor/new-value-store

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0025970-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Merge branch 'master' into refactor/new-value-store
